crates/fuel-core/src/service/sub_services/algorithm_updater.rs

```diff
@@ -290,31 +298,33 @@ fn sync_metadata_storage_with_on_chain_storage(
     Ok(())
 }
 
-fn sync_v0_metadata(
+fn sync_v0_metadata<L2DataStore, L2DataStoreView>(
     settings: &ConsensusParametersProvider,
-    on_chain_db: Database<OnChain, RegularStage<OnChain>>,
+    on_chain_db: &L2DataStoreView,
     metadata_height: u32,
     latest_block_height: u32,
     updater: &mut AlgorithmUpdaterV0,
     metadata_storage: &mut StructuredStorage<
         Database<GasPriceDatabase, RegularStage<GasPriceDatabase>>,
     >,
-) -> anyhow::Result<()> {
+) -> anyhow::Result<()>
+where
+    L2DataStore: L2Data,
+    L2DataStoreView: AtomicView<LatestView = L2DataStore>,
+{
     let first = metadata_height.saturating_add(1);
     let view = on_chain_db.latest_view()?;
     for height in first..=latest_block_height {
         let block = view
-            .storage::<FuelBlocks>()
-            .get(&height.into())?
+            .get_block(&height.into())?
             .ok_or(anyhow::anyhow!("Expected block to exist"))?;
         let last_tx_id = block.transactions().last().ok_or(anyhow::anyhow!(
             "Expected block to have at least one transaction"
         ))?;
         let param_version = block.header().consensus_parameters_version;
         let params = settings.settings(&param_version)?;
         let mint = view
```

Comment:
Hmm, the logic below `let block_gas_used = mint.mint_amount();` looks incorrect. In another places we calculate this value from several paramterers.

<img width="720" alt="image" src="https://github.com/user-attachments/assets/03b82792-cd74-4296-b68e-58d4895d7080">

I think we don't need ability to fetch the last transaction. I think we need to process the whole block as we do in the service itself. Sofrom port perspective we need only `latest_height` and `get_block` where `get_block` return whole block with all transactions.

It is an expensive operation than just fetching last transaction, but it will avoid potential issue and code will be easier to read, since ew use the same update logic during initialization, during `run` function and during `shutdown`.

Reply:
addressed in d48e3744b8, qq: (maybe @MitchTurner) what behaviour should we have for this line? - https://github.com/FuelLabs/fuel-core/blob/d48e3744b84b59dc400482d9fb88d04cfe89e9e0/crates/fuel-core/src/service/sub_services/algorithm_updater.rs#L332

I marked it as an error, since we have the `saturating_add` above.

Reply:
Yeah. That seems fine to me.